ui/app/components/timeline.ts

```diff
@@ -0,0 +1,94 @@
+import Component from '@glimmer/component';
+import { inject as service } from '@ember/service';
+import RouterService from '@ember/routing/router-service';
+import ApiService from 'waypoint/services/api';
+import { Build, Deployment, Release } from 'waypoint-pb';
+
+interface Artifact {
+  sequence: number;
+  type: string;
+  route: string;
+  completeTime: number;
+  isCurrentRoute: boolean;
+}
+
+interface TimelineArgs {
+  currentArtifact: Deployment.AsObject | Build.AsObject | Release.AsObject;
+}
+
+let TYPE_TRANSLATIONS = {
+  build: 'page.artifact.timeline.build',
+  deployment: 'page.artifact.timeline.deployment',
+  release: 'page.artifact.timeline.release',
+};
+
+let TYPE_ROUTES = {
+  build: 'workspace.projects.project.app.build',
+  deployment: 'workspace.projects.project.app.deployment',
+  release: 'workspace.projects.project.app.release',
+}
+
+export default class Timeline extends Component<TimelineArgs> {
+  @service api!: ApiService;
+  @service router!: RouterService;
+
+  get artifacts(): Artifact[] {
+    let timelineObjects = [] as Artifact[];
+    let preload = this.args.currentArtifact.preload;
+    let build = preload?.build ?? null;
+    let deployment = preload?.deployment ?? null;
```

Comment:
To get around the TypeScript errors here, I’d take a look at [type predicates](https://www.typescriptlang.org/docs/handbook/2/narrowing.html#using-type-predicates). Really powerful/useful feature of TypeScript.